_source/_posts/2019-11-19-secure-kafka-java.md

```diff
@@ -0,0 +1,458 @@
+```
+
+Access <http://localhost:8080/kafka/produce?message=This is my message> again to see the following message in the terminal running the Kafka consumer:
+
+```bash
+This is my message
+```
+
+Great job! You can stop this command for now. 
+
+Instead of executing from a terminal, let's add some Java code to consume the messages inside your app.
+
+## Consume Messages From a Kafka Topic in a Java App
+
+As with the producer, you need to add configurations to enable the consumer to find the Kafka Broker.
+
+Inside the `src/main/java/com/okta/javakafka/configuration` create the following class:
+
+```java
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
+import org.springframework.kafka.core.ConsumerFactory;
+import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Configuration
+public class ConsumerConfiguration {
+
+    private static final String KAFKA_BROKER = "localhost:9092";
+    private static final String GROUP_ID = "kafka-sandbox";
+
+    @Bean
+    public ConsumerFactory<String, String> consumerFactory() {
+        return new DefaultKafkaConsumerFactory<>(consumerConfigurations());
+    }
+
+    @Bean
+    public Map<String, Object> consumerConfigurations() {
+        Map<String, Object> configurations = new HashMap<>();
+
+        configurations.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, KAFKA_BROKER);
+        configurations.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
+        configurations.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        configurations.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+
+        return configurations;
+    }
+
+    @Bean
+    ConcurrentKafkaListenerContainerFactory<String, String> kafkaListenerContainerFactory() {
+        ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
+        factory.setConsumerFactory(consumerFactory());
+        return factory;
+    }
+
+}
+```
+
+The code above creates a factory that knows how to connect to your local broker. It also configures your consumer to deserialize a `String` for both the key and the value, matching the producer configuration.
+
+The Group ID is mandatory and used by Kafka to allow parallel data consumption. The `ConcurrentKafkaListenerContainerFactory` bean allows your app to consume messages in more than one thread.
+
+Now that your Java app is configured to find consumers inside your Kafka broker, let's start listening to the messages sent to the topic.
+
+Inside the `src/main/java/com/okta/javakafka/consumer` directory, create the following class:
```

Comment:
Need an instruction to create this directory

Reply:
Will do.